sympy/external/ntheory.py

```diff
@@ -153,3 +153,49 @@ def kronecker(x, y):
     if s % 2 and x % 8 in [3, 5]:
         sign = -sign
     return sign * jacobi(x, y)
+
+
+def iroot(y, n):
+    if y < 0:
+        raise ValueError("y must be nonnegative")
+    if n < 1:
+        raise ValueError("n must be positive")
+    if y in (0, 1):
+        return y, True
+    if n == 1:
+        return y, True
+    if n == 2:
+        x, rem = mlib.sqrtrem(y)
+        return int(x), not rem
+    if n >= y.bit_length():
+        return 1, False
+    # Get initial estimate for Newton's method. Care must be taken to
+    # avoid overflow
+    try:
+        guess = int(y**(1./n) + 0.5)
+    except OverflowError:
+        exp = math.log(y, 2)/n
+        if exp > 53:
+            shift = int(exp - 53)
+            guess = int(2.0**(exp - shift) + 1) << shift
+        else:
+            guess = int(2.0**exp)
```

Comment:
Are things like `2.0**exp` guaranteed not to overflow?

Let's see we have `n >= 3` so let's take `n = 3`.

We get here if `y ** 1./n` overflowed so `y > MAX_FLOAT ** 3`?

I have:
```
In [6]: sys.float_info.max
Out[6]: 1.7976931348623157e+308
```
So I need `y >= 10**1000` for `n = 3`.

Now `exp = log2(y)/n` so `2.0**exp ~= (y)**(1/n)`.

Somehow it seems like `shift` prevents this from overflowing but I can't quite convince myself that's true.

Reply:
Oh, of course it is because `shift ~= exp - 53` so `exp - shift ~= 53` and `2.0**53` won't overflow.